pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java

```diff
@@ -332,12 +332,17 @@ public void redeliverUnacknowledgedMessages(Consumer consumer, List<PositionImpl
     }
 
     @Override
-    protected void readMoreEntries(Consumer consumer) {
+    protected synchronized void readMoreEntries(Consumer consumer) {
         // consumer can be null when all consumers are disconnected from broker.
         // so skip reading more entries if currently there is no active consumer.
         if (null == consumer) {
             return;
         }
+        if (havePendingRead) {
```

Comment:
I see several parts of the code that call this `readMoreEntries` method after verifying that `havePendingRead` is `true`. I haven't yet verified the details of each call, but it seems like we should clean up those guards if we are moving the logic into this method.

Reply:
I think we should move the `havePendingRead` check to the `readMoreEntries ` method to avoid judgement everywhere.
Except for some judgments that have logical implications, I have changed them all.
Please take a look.